File: core/src/transcript.rs
use eyre::Result;
use num::integer::div_floor;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

pub fn format_timestamp(seconds: i64, always_include_hours: bool, decimal_marker: &str) -> String {
    assert!(seconds >= 0, "non-negative timestamp expected");
    let mut milliseconds = seconds * 10;

    let hours = div_floor(milliseconds, 3_600_000);
    milliseconds -= hours * 3_600_000;

    let minutes = div_floor(milliseconds, 60_000);
    milliseconds -= minutes * 60_000;

    let seconds = div_floor(milliseconds, 1_000);
    milliseconds -= seconds * 1_000;

    let hours_marker = if always_include_hours || hours != 0 {
        format!("{:02}:", hours)
    } else {
        String::new()
    };

    format!("{hours_marker}{minutes:02}:{seconds:02}{decimal_marker}{milliseconds:03}")
}

#[derive(Debug, Serialize, Deserialize, ToSchema, Clone)]
pub struct Transcript {
    pub processing_time_sec: u64,
    pub segments: Vec<Segment>,
}

#[derive(Debug, Serialize, Deserialize, ToSchema, Clone)]
pub struct Segment {
    pub start: i64,
    pub stop: i64,
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub speaker: Option<String>,
}

impl Segment {
    pub fn as_text(&self) -> String {
        self.text.to_owned()
    }

    pub fn as_vtt(&self) -> String {
        format!(
            "{} --> {}\n{}\n",
            format_timestamp(self.start, false, "."),
            format_timestamp(self.stop, false, "."),
            self.text.trim().replace("-->", "->")
        )
    }

    pub fn as_srt(&self, index: i32) -> String {
        format!(
            "\n{index}\n{} --> {}\n{}\n",
            format_timestamp(self.start, true, ","),
            format_timestamp(self.stop, true, ","),
            self.text.trim().replace("-->", "->")
        )
    }
}

impl Transcript {
    pub fn as_text(&self) -> String {
        self.segments
            .iter()
            .fold(String::new(), |transcript, fragment| transcript + fragment.text.as_str())
    }

    pub fn as_json(&self) -> Result<String> {
        Ok(serde_json::to_string_pretty(self)?)
    }

    pub fn as_vtt(&self) -> String {
        self.segments
            .iter()
            .fold(String::new(), |transcript, fragment| transcript + fragment.as_vtt().as_str())
    }

    pub fn as_srt(&self) -> String {
        self.segments
            .iter()
            .fold((1, String::new()), |(i, transcript), fragment| {
                (i + 1, transcript + fragment.as_srt(i).as_str())
            })
            .1
    }
}

File: desktop/src/lib/transcript.ts
export interface Duration {
	secs: number
	nanos: number
}

export interface Transcript {
	processing_time?: Duration
	segments: Segment[]
	word_segments?: Segment[]
}

export interface Segment {
	start: number
	stop: number
	text: string
	speaker?: number
}

/** Display names chosen for diarized speakers, keyed by the segment's zero-based speaker index. */
export type SpeakerNames = Record<number, string>

/** What a speaker is called on screen and in exports: the chosen name, else "Speaker N". */
export function speakerName(speaker: number, label: string, names?: SpeakerNames): string {
	const chosen = names?.[speaker]?.trim()
	return chosen || `${label} ${speaker + 1}`
}

export function formatTimestamp(seconds: number, alwaysIncludeHours: boolean, decimalMarker: string, includeMilliseconds: boolean = true): string {
	if (seconds < 0) {
		throw new Error('Non-negative timestamp expected')
	}

	let milliseconds = seconds * 10

	const hours = Math.floor(milliseconds / 3_600_000)
	milliseconds -= hours * 3_600_000

	const minutes = Math.floor(milliseconds / 60_000)
	milliseconds -= minutes * 60_000

	const formattedSeconds = Math.floor(milliseconds / 1_000)
	milliseconds -= formattedSeconds * 1_000

	const hoursMarker = alwaysIncludeHours || hours !== 0 ? `${String(hours).padStart(2, '0')}:` : ''

	let result = `${hoursMarker}${String(minutes).padStart(2, '0')}:${String(formattedSeconds).padStart(2, '0')}`

	if (includeMilliseconds) {
		result += `${decimalMarker}${String(milliseconds).padStart(3, '0')}`
	}

	return result
}

function speakerPrefix(segment: Segment, label: string, names?: SpeakerNames): string {
	return segment.speaker != null ? `[${speakerName(segment.speaker, label, names)}] ` : ''
}

export function asSrt(segments: Segment[], speakerLabel: string = 'Speaker', speakerNames?: SpeakerNames) {
	return segments.reduce((transcript, segment, i) => {
		return (
			transcript +
			`${i > 0 ? '\n' : ''}${i + 1}\n` +
			`${formatTimestamp(segment.start, true, ',')} --> ${formatTimestamp(segment.stop, true, ',')}\n` +
			`${speakerPrefix(segment, speakerLabel, speakerNames)}${segment.text.trim().replace('-->', '->')}\n`
		)
	}, '')
}

export function asVtt(segments: Segment[], speakerLabel: string = 'Speaker', speakerNames?: SpeakerNames) {
	return segments.reduce((transcript, segment) => {
		return (
			transcript +
			`${formatTimestamp(segment.start, false, '.')} --> ${formatTimestamp(segment.stop, false, '.')}\n` +
			`${speakerPrefix(segment, speakerLabel, speakerNames)}${segment.text.trim().replace('-->', '->')}\n`
		)
	}, '')
}

export function asText(segments: Segment[], speakerLabel: string = 'Speaker', speakerNames?: SpeakerNames) {
	return segments.reduce((transcript, segment) => {
		return transcript + `${speakerPrefix(segment, speakerLabel, speakerNames)}${segment.text.trim()}\n`
	}, '')
}

export function normalizeWhitespace(text: string) {
	return text.replace(/\s+/g, ' ').trim()
}

export function asJson(segments: Segment[]) {
	return JSON.stringify(
		segments.map((s) => ({
			...s,
			start: s.start / 100,
			stop: s.stop / 100,
		})),
		null,
		4,
	)
}

function escapeCsv(value: string) {
	return `"${value.replace(/"/g, '""')}"`
}

export function asCsv(segments: Segment[]) {
	const hasSpeakers = segments.some((s) => s.speaker != null)
	const header = hasSpeakers ? 'start,end,speaker,text' : 'start,end,text'
	const rows = segments.map((segment) => {
		const start = formatTimestamp(segment.start, true, '.')
		const end = formatTimestamp(segment.stop, true, '.')
		const text = segment.text.trim()
		if (hasSpeakers) {
			const speaker = segment.speaker != null ? String(segment.speaker + 1) : ''
			return `${escapeCsv(start)},${escapeCsv(end)},${escapeCsv(speaker)},${escapeCsv(text)}`
		}
		return `${escapeCsv(start)},${escapeCsv(end)},${escapeCsv(text)}`
	})
	return [header, ...rows].join('\n')
}
